[node-e2e]: Support checking if a feature gate is enabled during functional tests

iholder101 opened the pull request:
#### What type of PR is this?
/kind feature
/kind flake

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Today it isn't possible to check whether a feature gate is enabled or not during node e2e testing. Such a capability is important in certain cases, for example for [swap tests](https://github.com/kubernetes/kubernetes/blob/v1.29.0-alpha.0/test/e2e_node/swap_test.go#L145) [1]. However, the current way to check if this feature gate is enabled does not work and always returns false.

This PR introduces an `isFeatureGateEnabled()` function to determine if a feature gate is enabled or not in a reliable manner. For more info read below.

[1] https://github.com/kubernetes/kubernetes/blob/v1.29.0-alpha.0/test/e2e_node/swap_test.go#L145

#### Special notes for your reviewer:
Turns out that the `skipper` package introduced this very logic [here](https://github.com/kubernetes/kubernetes/blob/v1.29.0-alpha.0/test/e2e/framework/skipper/skipper.go#L61) [1], explaining that it `must be called in test suites that have a --feature-gates parameter`. However, the `featureGate` object, although [initialized](https://github.com/kubernetes/kubernetes/blob/v1.29.0-alpha.0/test/e2e_node/e2e_node_suite_test.go#L138) [2], is never actually used.

What I did in this PR is introduce this object in the `e2enode` package, initialized it, and created an `isFeatureGateEnabled()` function that uses it. Then, I've cleaned up the unused code from the skipper package.

[1] https://github.com/kubernetes/kubernetes/blob/v1.29.0-alpha.0/test/e2e/framework/skipper/skipper.go#L61
[2] https://github.com/kubernetes/kubernetes/blob/v1.29.0-alpha.0/test/e2e_node/e2e_node_suite_test.go#L138

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

iholder101 committed:
[node-e2e]: Init a FeatureGate interface to check if a FG is enabled

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Remove undeeded featureGate object from skipper package

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
/sig node

harche commented:
/test pull-kubernetes-node-e2e-containerd
/test pull-kubernetes-node-crio-e2e

iholder101 commented:
/test pull-kubernetes-e2e-kind

mmiranda96 commented:
/assign @harche

iholder101 commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1

/test pull-kubernetes-node-swap-ubuntu-serial

/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial

iholder101 commented:
/retest

harche commented:
/lgtm

mmiranda96 commented:
/lgtm
/retest

pacoxu commented:
```
E2eNode Suite: [It] [sig-node] Swap [NodeConformance][LinuxOnly] with configuration QOS Burstable

Reason: skipped
```
Swap tests are still skipped in `pull-kubernetes-node-swap-fedora-serial` and `pull-kubernetes-node-swap-ubuntu-serial. 
`pull-kubernetes-node-swap-fedora` will run the swap e2e tet. Is this expected?

pacoxu commented:
/cc @pohly

pacoxu commented:
/priority important-soon
/triage accepted

pohly commented on test/e2e_node/util.go:
```diff
@@ -91,9 +91,29 @@ var (
 	// KubeletConfig is the kubelet configuration the test is running against.
-	kubeletCfg *kubeletconfig.KubeletConfiguration
+	kubeletCfg  *kubeletconfig.KubeletConfiguration
+	featureGate featuregate.FeatureGate
```
Is this variable really needed? Can't the `utilfeature.DefaultMutableFeatureGate` be used instead?

iholder101 replied:
Hey @pohly! Thanks for reviewing!

That's a good question.
The previous function from `skipper` package was cloning `utilfeature.DefaultMutableFeatureGate`.

Do you think there's a good reason to avoid mutating `utilfeature.DefaultMutableFeatureGate` directly? If not I'd be happy to get rid of the `featureGate` variable.

pohly replied:
A package should not mutate utilfeature.DefaultMutableFeatureGate, but `test/e2e_node` is the "main binary" and can modify it.

There is also an `AddFlag` method, but I think that needs pflag and thus can't be used here.

pohly commented on test/e2e/framework/skipper/skipper.go:
```diff
@@ -56,50 +55,6 @@ func SkipUnlessAtLeast(value int, minValue int, message string) {
-		skipInternalf(1, "Only supported when %v feature is disabled", gate)
-	}
-}
-
```
If `e2e_node` was the only potential user of this code and even that doesn't really use the "skipping" aspect of it, then I agree that this code should get removed.

My last cleanup was an attempt to catch incorrect usage of `SkipUnlessFeatureGateEnabled/SkipIfFeatureGateEnabled`, but that obviously didn't help when the code wasn't even getting called at all.

iholder101 replied:
:)
Yeah I was pretty surprised that these functions aren't being called, perhaps they were in the past.

As a side note: IMHO the best way of skipping tests is from the outside rather than in-test skips, since it's sometimes easy to miss a test being skipped.

pohly replied:
They were used incorrectly in the past, which I fixed at least in some places: 12990dec4057847bf6e7e8037c27baa16e58648b

SergeyKanzhelev replied:
I think all tests needed it GA-ed. We will have a need for it again at some point. I am fine with removing it for now.

Long term the reliable way to test feature gates may be to query /config endpoint of the kubelet.

aojea replied:
this was a source of confusion, if we can remove it the better, long term we should build something better documented and consistent

bart0sh replied:
I'm using `SkipUnlessFeatureGateEnabled` in the test case I'm developing now. It tests CDI injections in the e2e_node/device_plugin_test.go. The test case should be skipped if the future gate `DevicePluginCDIDevices` is not enabled.

What should I use instead?

iholder101 commented:
> ```
> E2eNode Suite: [It] [sig-node] Swap [NodeConformance][LinuxOnly] with configuration QOS Burstable
> 
> Reason: skipped
> ```
> 
> Swap tests are still skipped in `pull-kubernetes-node-swap-fedora-serial` and `pull-kubernetes-node-swap-ubuntu-serial. `pull-kubernetes-node-swap-fedora` will run the swap e2e tet. Is this expected?

Thanks @pacoxu!

Yes, this is expected, as the current swap tests run as part of `NodeConformance`, which is being skipped by the lanes you specified. Thanks for raising that, on a second thought I shouldn't have run these lanes in this PR. Sorry for the confusion.

iholder101 committed:
Remove featureGates variable and initFeatureGates() function

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
> Is this variable really needed? Can't the `utilfeature.DefaultMutableFeatureGate` be used instead?

@pohly I've removed the featureGate variable like we discussed.

@pacoxu @harche PTAL

iholder101 commented:
@pohly Can you please look at the [failing lane](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/120139/pull-kubernetes-verify/1696122271350919168) here? The error is:

```
Test files may not access mutable global feature gates directly:
test/e2e_node/e2e_node_suite_test.go:	if err := utilfeature.DefaultMutableFeatureGate.SetFromMap(featureGates); err != nil {

Use this invocation instead:
  defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, <value>)()
```

It comes from here:
https://github.com/kubernetes/kubernetes/blob/7e0a9a7f4e70e4cc6f5dd49917d18101a8e1a25c/hack/verify-test-featuregates.sh#L32-L42

To me it doesn't seem like a problem, since as you said this isn't a regular "test file".
One way of solving this would be to rename the file and remove the `_test` postfix. I think it kind of makes sense since if it's not really a test.

WDYT?

pohly commented:
> > Test files may not access mutable global feature gates directly:

Yes, this is a false positive.

> One way of solving this would be to rename the file and remove the _test postfix.

It might be that Go only supports `TestE2eNode` as test function when it is in a `_test.go` file.

A less intrusive alternative is to move only the feature gate initialization into a non-test file (e.g. `features.go`).

iholder101 committed:
Move feature gate initialization to util.go

Without this commit the following error is raised since
*_test.go aren't allowed to mutate global feature gates:

Test files may not access mutable global feature gates directly:
test/e2e_node/e2e_node_suite_test.go:	if err := utilfeature.DefaultMutableFeatureGate.SetFromMap(featureGates); err != nil {

Use this invocation instead:
  defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, <value>)()

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
> > > Test files may not access mutable global feature gates directly:
> 
> Yes, this is a false positive.
> 
> > One way of solving this would be to rename the file and remove the _test postfix.
> 
> It might be that Go only supports `TestE2eNode` as test function when it is in a `_test.go` file.
> 
> A less intrusive alternative is to move only the feature gate initialization into a non-test file (e.g. `features.go`).

@pohly Done! Moved to `util.go`, linter is quite now :)

@mmiranda96 @harche
PTAL

harche commented:
/retest-required

SergeyKanzhelev commented on test/e2e_node/util.go:
```diff
@@ -94,6 +94,19 @@ var (
 	kubeletCfg *kubeletconfig.KubeletConfiguration
 )
 
+func initFeatureGates(featureGates map[string]bool) {
```
there is also `serviceFeatureGates`. Can we initialize both without the globals?

SergeyKanzhelev replied:
or rename another method to highlight that it checks kubelet gates, not service gates

iholder101 replied:
> there is also serviceFeatureGates. Can we initialize both without the globals?

Yes, looking at it again it seems that `initFeatureGates()` and `SetFeatureGatesForInProcessComponents` basically do the same thing, which is essentially `DefaultMutableFeatureGate.SetFromMap(featureGates)`.

I'm not sure I understand your meaning behind "without the globals". Do you mean that we should have another `MutableFeatureGate` variable (similarly to `DefaultMutableFeatureGate`) that would be dedicated to service gates? Something like:

```go
var (
	DefaultMutableFeatureGate featuregate.MutableFeatureGate = featuregate.NewFeatureGate()
	DefaultMutableServiceGate featuregate.MutableFeatureGate = featuregate.NewFeatureGate()
)
```

Then I can modify `initFeatureGates()` to get the global feature gate variable as an argument, and do something like:
```go
initFeatureGates(featureGates, DefaultMutableFeatureGate)
initFeatureGates(serviceFeatureGates, DefaultMutableServiceGate)
```

We can also create separate methods to check for a feature gate enablement and a service gate enablement.
Was this your meaning? Please let me know if I didn't understand you correctly.

iholder101 replied:
Also, I wonder what are service gates exactly and how they differ from "regular" feature gates

SergeyKanzhelev replied:
we are controlling fake API server feature gates separately from kubelet feature gates. See the use of `serviceFeatureGates` variable

iholder101 replied:
@SergeyKanzhelev Done, PTAL at the last commit, I hope it aligns with your expectations.

Small question, WDYT about:
https://github.com/kubernetes/kubernetes/blob/7e0a9a7f4e70e4cc6f5dd49917d18101a8e1a25c/test/e2e_node/e2e_node_suite_test.go#L230-L237

Should this code use `serviceFeatureGates` instead of `featureGates`?

iholder101 replied:
Unfortunately the linter lane [fails](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/120139/pull-kubernetes-linter-hints/1702230123148742656) since the new `isServiceFeatureGateEnabled()` function is not used =/

What do you suggest? Should we remove it for now?

SergeyKanzhelev commented on test/e2e_node/util.go:
```diff
@@ -94,6 +94,19 @@ var (
+	}
+}
+
+func isFeatureGateEnabled(feature featuregate.Feature) bool {
```
let's add a comment saying that those are feature gates set at test start. I vagualy remember people was trying to add tests that will update feature gates and restart kubelet to validate the use case when feature gate is being disabled

SergeyKanzhelev commented on staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go:
```diff
@@ -27,6 +27,9 @@ var (
 	DefaultMutableFeatureGate featuregate.MutableFeatureGate = featuregate.NewFeatureGate()
 
+	// DefaultMutableServiceGate is similar to DefaultMutableFeatureGate, but is intended for API server feature gates.
+	DefaultMutableServiceGate featuregate.MutableFeatureGate = featuregate.NewFeatureGate()
```
sorry, I think we are talking past each other. This file is being used to set the process-wide feature gates. So we only need to keep `DefaultMutableFeatureGate`.

The e2e_node process can be configured with two arguments: one defines feature gates that will be passed to  the kubelet process that test will start. Another argument defines feature gates to set for the fake API server that will be served by the test.

The fact there are two different arguments for e2e_node is a specific of test configuration and should not leak into the common library like this.

iholder101 replied:
Thanks for the clarification. I've now moved `DefaultMutableServiceGate` to `test/e2e_node/util.go` and made it unexposed (i.e. "private").

Does this fulfills your expectations? If it doesn't, please kindly clarify your expectations.

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

iholder101 committed:
Separate between feature gates and service gates

With this commit we would have two sets of feature gates,
kubelet feature gates and API server feature gates.
There are also two dedicated funtions to check
a feature / service gate enablement.

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
/test pull-kubernetes-typecheck

iholder101 commented:
/test pull-kubernetes-node-e2e-containerd

SergeyKanzhelev commented:
/lgtm

iholder101 commented:
@SergeyKanzhelev linter is angry again..
```
> hack/verify-test-featuregates.sh
Test files may not access mutable global feature gates directly:
test/e2e_node/e2e_node_suite_test.go:	initFeatureGates(utilfeature.DefaultMutableFeatureGate, featureGates)

Use this invocation instead:
  defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, <value>)()
```

Is there a way to exclude these lines from the linter? I understand its general reasoning, but this is not just a regular test file. WDYT?

mmiranda96 commented:
[Linter](https://github.com/kubernetes/kubernetes/blob/cde6a46a483a0cc7d3f2404db506e1c24ecd4e9d/hack/verify-test-featuregates.sh#L33) seems to focus only on `_test.go` files, and the only file having this issue is `test/e2e_node/e2e_node_suite_test.go `. I think excluding that file sounds reasonable, but let's add a comment explaining why.

dims commented:
cc @pohly

pohly commented on test/e2e_node/e2e_node_suite_test.go:
```diff
@@ -135,15 +133,9 @@ func TestMain(m *testing.M) {
-		os.Exit(1)
-	}
+	initFeatureGates(utilfeature.DefaultMutableFeatureGate, featureGates)
+	initFeatureGates(defaultMutableServiceGate, serviceFeatureGates)
```
Can't you move this into a non-test file, like it was in some previous revision of this PR? Then we don't need to add exceptions to the verify script.

iholder101 replied:
Done!
linter is happy now :)

Thank you and @mmiranda96

iholder101 committed:
[Refactor] global feature gate objects to be set in a non _test file

The linter does not allot _test.go files to set
global feature gate objects.

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
PTAL @pohly

pohly commented on test/e2e_node/swap_test.go:
```diff
@@ -144,7 +143,7 @@ func runPodAndWaitUntilScheduled(f *framework.Framework, pod *v1.Pod) *v1.Pod {
 func isSwapFeatureGateEnabled() bool {
 	ginkgo.By("figuring if NodeSwap feature gate is turned on")
-	return utilfeature.DefaultFeatureGate.Enabled(features.NodeSwap)
+	return isFeatureGateEnabled(features.NodeSwap)
```
Can't you use `utilfeature.DefaultFeatureGate.Enabled(feature)` here and remove the `isFeatureGateEnabled` function? The new comments there can be moved to this function.

Having two functions which are both just thin wrappers makes it harder to figure out what the code actually does.

iholder101 replied:
Yeah, I agree. Thanks!
Done.

pohly commented on test/e2e_node/util.go:
```diff
@@ -75,6 +75,9 @@ var startServices = flag.Bool("start-services", true, "If true, start local node
 var busyboxImage = imageutils.GetE2EImage(imageutils.BusyBox)
 
+// defaultMutableServiceGate is similar to DefaultMutableFeatureGate, but is intended for API server feature gates.
+var defaultMutableServiceGate featuregate.MutableFeatureGate = featuregate.NewFeatureGate()
```
Why is is called `ServiceGate` instead of `FeatureGate`?

And why `default`?

What is this used for? If it's not used, remove it.

iholder101 replied:
Please see Sergey's comments above [[1](https://github.com/kubernetes/kubernetes/pull/120139#discussion_r1326349635)][[2](https://github.com/kubernetes/kubernetes/pull/120139#discussion_r1323685079)].

In short, as @SergeyKanzhelev said, the `serviceFeatureGates` argument "defines feature gates to set for the fake API server that will be served by the test".

> What is this used for?

To check whether service feature gates are enabled or not during tests.

> Why is is called ServiceGate instead of FeatureGate? And why default?

We can discuss the name. My reasoning was to follow `DefaultMutableFeatureGate`'s naming scheme, but for service feature gates. Here are some alternatives I can think about:
- mutableServiceGate
- defualtServiceGate
- serviceFeatureGate

Let me know if you have other suggestions.

> If it's not used, remove it.

Before this PR, checking whether a feature gate is enabled during tests was impossible, or at least extremely non-intuitive. Using `DefaultFeatureGate.Enabled()` did not work and always returned false since the variable was never properly initialized.

IMHO if a day comes where there would be a need to check if a service feature gate is enabled during tests it would be non-intuitive to figure out how to do so.

WDYT?

dchen1107 replied:
+1 @pohly. cc @SergeyKanzhelev

iholder101 committed:
Add documentation to DefaultMutableFeatureGate regarding use in tests

The documentation clarifies that this variable can be used only for
feature gates that are being enabled at start test time, and not
for feature gates that are being enabled during the test.

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
/test pull-kubernetes-node-e2e-containerd

aojea commented:
it will be better to merge it with a test that exercises it (and also work as example for new tests implementers) , otherwise we have the risk of having yet another mechanism that may be orphan

iholder101 commented:
> it will be better to merge it with a test that exercises it (and also work as example for new tests implementers) , otherwise we have the risk of having yet another mechanism that may be orphan

A test already uses it, for example: https://github.com/kubernetes/kubernetes/blob/v1.29.0-alpha.1/test/e2e_node/swap_test.go#L147.

Furthermore, we've notice failures lately because the mechanism does not work properly as of now. This PR fixes it.

pacoxu commented:
/retest

pacoxu commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1

/test pull-kubernetes-node-swap-ubuntu-serial

/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial

pacoxu commented:
@pohly @aojea do you have time to take a look again? This hangs the swap e2e test.

pacoxu commented:
/lgtm

pohly commented on staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go:
```diff
@@ -25,6 +25,9 @@ var (
 	//   defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, <value>)()
+	//
+	// When used in tests, it must be used for feature gates that are enabled at test start time.
+	// It will not work if the feature gate is being enabled during the test.
```
"Will not work" is a pretty strong statement. I suppose you ran into a situation where some test startup code was checking a feature gate, did something based on it, and later the test changed the gate.

But there might also be tests where it's okay because the code that is depending on the feature runs later or simply always checks the feature without any lasting effect.

I think I prefer to not add this comment.

bart0sh commented:
@iholder101 
> Today it isn't possible to check whether a feature gate is enabled or not during node e2e testing.

FWIW, it worked for me just fine with `--service-feature-gates` command line option:
```
sudo _output/bin/e2e_node.test --service-feature-gates="DevicePluginCDIDevices=true" --ginkgo.v --ginkgo.focus='Can schedule a pod that requires a device'
...
STEP: >> DevicePluginCDIDevices feature gate is enabled: true @ 10/22/23 18:16:27.916
...
```

and this change in the test code:
```
diff --git a/test/e2e_node/device_plugin_test.go b/test/e2e_node/device_plugin_test.go
index de3c7f56556..1ee6f2e7125 100644
--- a/test/e2e_node/device_plugin_test.go
+++ b/test/e2e_node/device_plugin_test.go
@@ -181,6 +181,7 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
                })
 
                ginkgo.It("Can schedule a pod that requires a device", func(ctx context.Context) {
+                       ginkgo.By(fmt.Sprintf(">> DevicePluginCDIDevices feature gate is enabled: %t", utilfeature.DefaultFeatureGate.Enabled(features.DevicePluginCDIDevices)))
                        podRECMD := fmt.Sprintf("devs=$(ls /tmp/ | egrep '^Dev-[0-9]+$') && echo stub devices: $devs && sleep %s", sleepIntervalWithRestart)
                        pod1 := e2epod.NewPodClient(f).CreateSync(ctx, makeBusyboxPod(SampleDeviceResourceName, podRECMD))
                        deviceIDRE := "stub devices: (Dev-[0-9]+)"

```

It's a bit tricky, but not impossible.

bart0sh commented:
/cc

kannon92 commented:
> @iholder101
> 
> > Today it isn't possible to check whether a feature gate is enabled or not during node e2e testing.
> 
> FWIW, it worked for me just fine with `--service-feature-gates` command line option:
> 
> ```
> sudo _output/bin/e2e_node.test --service-feature-gates="DevicePluginCDIDevices=true" --ginkgo.v --ginkgo.focus='Can schedule a pod that requires a device'
> ...
> STEP: >> DevicePluginCDIDevices feature gate is enabled: true @ 10/22/23 18:16:27.916
> ...
> ```
> 
> and this change in the test code:
> 
> ```
> diff --git a/test/e2e_node/device_plugin_test.go b/test/e2e_node/device_plugin_test.go
> index de3c7f56556..1ee6f2e7125 100644
> --- a/test/e2e_node/device_plugin_test.go
> +++ b/test/e2e_node/device_plugin_test.go
> @@ -181,6 +181,7 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
>                 })
>  
>                 ginkgo.It("Can schedule a pod that requires a device", func(ctx context.Context) {
> +                       ginkgo.By(fmt.Sprintf(">> DevicePluginCDIDevices feature gate is enabled: %t", utilfeature.DefaultFeatureGate.Enabled(features.DevicePluginCDIDevices)))
>                         podRECMD := fmt.Sprintf("devs=$(ls /tmp/ | egrep '^Dev-[0-9]+$') && echo stub devices: $devs && sleep %s", sleepIntervalWithRestart)
>                         pod1 := e2epod.NewPodClient(f).CreateSync(ctx, makeBusyboxPod(SampleDeviceResourceName, podRECMD))
>                         deviceIDRE := "stub devices: (Dev-[0-9]+)"
> ```
> 
> It's a bit tricky, but not impossible.

Thanks @bart0sh.

I was able to confirm that we can do this locally.  

Have https://github.com/kubernetes/test-infra/pull/31134 to add service-feature-gates to the swap tests.

bart0sh commented:
> Have https://github.com/kubernetes/test-infra/pull/31134 to add service-feature-gates to the swap tests.

In my opinion it would be better to fix `utilfeature.DefaultFeatureGate.Enabled`. It should return `true` for e2e_node tests if `--feature-gates` is used and shouldn't require `----service-feature-gates`.

bart0sh commented:
@iholder101 are you going to continue working on this? If so, please rebase, fix failing CI and address review comments, thanks.

harche commented:
> @iholder101 are you going to continue working on this? If so, please rebase, fix failing CI and address review comments, thanks.

@iholder101 is not available due to unforeseen circumstances, but I believe @kannon92 has already addressed the issue in some other PR. 

But considering the unavailability of @iholder101, I think would be okay to close this issue. 

/close

pacoxu commented:
> @iholder101 is not available due to unforeseen circumstances, but I believe @kannon92 has already addressed the issue in some other PR.

Do you mean https://github.com/kubernetes/kubernetes/pull/122175/files#r1416359942 or https://github.com/kubernetes/kubernetes/pull/122745/files#diff-40c4fc8e2327888a122bfcb2f8e7232e0164639a3f2da621dd4d9a70b1eb1ee4R248-R253?

kannon92 commented:
So IMO our tests don't really support this. I decided to go a different route and just turn on the feature for these jobs.

In those PRs, I ended up modifying test-infra to turn these on (using feature-gate and service-feature-gate) and I got rid of the check as we know that it is turned on via the test setup.

kannon92 commented:
My goal was to get these swap tests working without having to introduce a new way. I think this idea is useful but wanted to decouple this from the failing swap tests.
